eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -145,11 +196,51 @@ final case class RES_GETINFO(nodeId: PublicKey, channelId: ByteVector32, state:
-  case class ChannelOpened(channelId: ByteVector32) extends ChannelCommandResponse { override def toString  = s"created channel $channelId" }
-  case class ChannelClosed(channelId: ByteVector32) extends ChannelCommandResponse { override def toString  = s"closed channel $channelId" }
+/** response to [[Command]] requests */
+sealed trait CommandResponse[+C <: Command]
+sealed trait CommandSuccess[+C <: Command] extends CommandResponse[C]
+sealed trait CommandFailure[+C <: Command, +T <: Throwable] extends CommandResponse[C] { def t: Throwable }
+
+/** generic responses */
+final case class RES_SUCCESS[+C <: Command](cmd: C, channelId: ByteVector32) extends CommandSuccess[C]
+final case class RES_FAILURE[+C <: Command, +T <: Throwable](cmd: C, t: T) extends CommandFailure[C, T]
+
+/**
+ * special case for [[CMD_ADD_HTLC]]
+ * note that for this command there is gonna be two response patterns:
+ * - either [[RES_ADD_FAILED]]
+ * - or [[RES_SUCCESS[CMD_ADD_HTLC]]] followed by [[RES_ADD_COMPLETED]] (possibly a while later)
+ */
+final case class RES_ADD_FAILED[+T <: Throwable](c: CMD_ADD_HTLC, t: T, channelUpdate: Option[ChannelUpdate]) extends CommandFailure[CMD_ADD_HTLC, T] { override def toString = s"cannot add htlc with origin=${c.origin} reason=${t.getMessage}" }
+sealed trait HtlcResult
+object HtlcResult {
+  sealed trait Fulfill extends HtlcResult { def paymentPreimage: ByteVector32 }
+  case class RemoteFulfill(fulfill: UpdateFulfillHtlc) extends Fulfill { override val paymentPreimage = fulfill.paymentPreimage }
+  case class OnChainFulfill(paymentPreimage: ByteVector32) extends Fulfill
+  sealed trait Fail extends HtlcResult
+  case class RemoteFail(fail: UpdateFailHtlc) extends Fail
+  case class RemoteFailMalformed(fail: UpdateFailMalformedHtlc) extends Fail
+  case class OnChainFail(cause: ChannelException) extends Fail
+  case class Disconnected(channelUpdate: ChannelUpdate) extends Fail { assert(!Announcements.isEnabled(channelUpdate.channelFlags), "channel update must have disabled flag set") }
+}
+final case class RES_ADD_COMPLETED[+O <: Origin[_ <: Upstream], +R <: HtlcResult](to: O, htlc: UpdateAddHtlc, result: R) extends CommandSuccess[CMD_ADD_HTLC]
+
+/** other specific responses */
+final case class RES_GETSTATE[+S <: State](state: S) extends CommandSuccess[CMD_GETSTATE.type]
+final case class RES_GETSTATEDATA[+D <: Data](data: D) extends CommandSuccess[CMD_GETSTATEDATA.type]
+final case class RES_GETINFO(nodeId: PublicKey, channelId: ByteVector32, state: State, data: Data) extends CommandSuccess[CMD_GETINFO.type]
+final case class RES_CLOSE(channelId: ByteVector32) extends CommandSuccess[CMD_CLOSE]
+
+/**
+ * Those are not response to [[Command]], but to [[fr.acinq.eclair.io.Peer.OpenChannel]]
+ *
+ * If actor A sends a [[fr.acinq.eclair.io.Peer.OpenChannel]] and actor B sends a [[CMD_CLOSE]], then A will receive a
+ * [[ChannelOpenResponse.ChannelClosed]] whereas B will receive a [[]]
```

Comment:
What B will receive seems to be missing...it's a `[[RES_CLOSE]]` IIUC

Reply:
Done with c46fff704.